File: 209.长度最小的子数组.py
#
# @lc app=leetcode.cn id=209 lang=python3
#
# [209] 长度最小的子数组
#

# @lc code=start
class Solution:
    def minSubArrayLen(self, s: int, nums: List[int]) -> int:
        n = len(nums)
        if n == 0:
            return 0

        INT_MAX = 999999999
        ans = INT_MAX
        start = end = 0
        sums = 0 
        while end < n:
            sums += nums[end]
            while sums >= s:
                ans = min(ans, end - start + 1)
                sums -= nums[start]
                start += 1
            end += 1
        ans = 0 if ans == INT_MAX else ans
        
        return ans 
# @lc code=end

File: 209.长度最小的子数组.cpp
/*
 * @lc app=leetcode.cn id=209 lang=cpp
 *
 * [209] 长度最小的子数组
 */


// @lc code=start
class Solution {
public:
    // O(n^2)
    // int minSubArrayLen(int s, vector<int>& nums) {
    //     int n = nums.size();
    //     if (n == 0) {
    //         return 0;
    //     }
    //     int ans = INT_MAX;
    //     for (int i = 0; i < n; i++)
    //     {
    //         int sum = 0;
    //         for (int j = i; j < n; j++)
    //         {
    //             sum += nums[j];
    //             if (sum >= s)
    //             {
    //                 ans = min(ans, j - i + 1);
    //                 break;
    //             }
    //         }
    //     }
    //     return ans == INT_MAX ? 0 : ans;
    // }
    
    // O(n): 双指针，其中 n 是数组的长度。指针start和end最多各移动 n 次
    int minSubArrayLen(int s, vector<int>& nums) {
        int n = nums.size();
        if (n == 0)
        {
            return 0;
        }

        int MAX_VALUE = INT_MAX;
        int ans = MAX_VALUE;

        int start = 0, end = 0;   // 初始化
        int sum = 0;
        
        while (end < n)
        {
            sum += nums[end];
            while (sum >= s)
            {
                ans = min(ans, end - start + 1);
                sum -= nums[start];
                start++;

            }
            end++;
        }
        
        return ans == MAX_VALUE ? 0 : ans;
      
    }

};
// @lc code=end